executor/aggregate.go

```diff
@@ -816,6 +853,117 @@ func (e *HashAggExec) getPartialResults(groupKey string) []aggfuncs.PartialResul
 	return partialResults
 }
 
+func (e *HashAggExec) initRuntimeStats() {
+	if e.runtimeStats != nil && e.stats == nil {
+		e.stats = &HashAggRuntimeStats{
+			PartialNum:        e.ctx.GetSessionVars().HashAggPartialConcurrency(),
+			FinalNum:          e.ctx.GetSessionVars().HashAggFinalConcurrency(),
+			PartialStats:      &AggWorkerStat{},
+			FinalStats:        &AggWorkerStat{},
+			PartialWorkerTime: make([]time.Duration, e.ctx.GetSessionVars().HashAggPartialConcurrency()),
+			FinalWorkerTime:   make([]time.Duration, e.ctx.GetSessionVars().HashAggFinalConcurrency()),
+		}
+		e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.RegisterStats(e.id, e.stats)
+	}
+}
+
+// HashAggRuntimeStats record the HashAggExec runtime stat
+type HashAggRuntimeStats struct {
+	PartialNum int
+	FinalNum   int
+
+	PartialStats *AggWorkerStat
+	FinalStats   *AggWorkerStat
+
+	PartialWorkerTime []time.Duration
+	FinalWorkerTime   []time.Duration
+}
+
+// AggWorkerStat record the AggWorker runtime stat
+type AggWorkerStat struct {
+	TaskNum  int64
+	WallTime time.Duration
+	WaitTime int64
+	ExecTime int64
+}
+
+func (e *HashAggRuntimeStats) String() string {
+	var result string
+	if e.PartialStats.WallTime != 0 {
+		var totalTime time.Duration = 0
+		for _, v := range e.PartialWorkerTime {
+			totalTime += v
+		}
+		result += fmt.Sprintf("partial_worker:{wall_time:%s, concurrency:%d, task_num:%d, tot_wait:%s, tot_exec:%s, tot_time:%s", e.PartialStats.WallTime, e.PartialNum, e.PartialStats.TaskNum, time.Duration(e.PartialStats.WaitTime), time.Duration(e.PartialStats.ExecTime), totalTime)
+	}
+	if e.PartialWorkerTime != nil {
+		sort.Slice(e.PartialWorkerTime, func(i, j int) bool { return e.PartialWorkerTime[i] < e.PartialWorkerTime[j] })
+		n := len(e.PartialWorkerTime)
+		if n != 0 {
+			result += fmt.Sprintf(", max:%v, p95:%v", e.PartialWorkerTime[n-1], e.PartialWorkerTime[n*19/20])
+		}
+	}
+	if len(result) > 0 {
+		result += "}, "
+	}
+	if e.FinalStats.WallTime != 0 {
+		var finalTotal time.Duration = 0
+		for _, v := range e.FinalWorkerTime {
+			finalTotal += v
+		}
+		result += fmt.Sprintf("final_worker:{wall_time:%s, concurrency:%d, task_num:%d, tot_wait:%s, tot_exec:%s, tot_time:%s", e.FinalStats.WallTime, e.FinalNum, e.FinalStats.TaskNum, time.Duration(e.PartialStats.WaitTime), time.Duration(e.PartialStats.ExecTime), finalTotal)
+	}
+	if e.FinalWorkerTime != nil {
+		sort.Slice(e.FinalWorkerTime, func(i, j int) bool { return e.FinalWorkerTime[i] < e.FinalWorkerTime[j] })
+		m := len(e.FinalWorkerTime)
+		if m != 0 {
+			result += fmt.Sprintf(", max:%v, p95:%v", e.FinalWorkerTime[m-1], e.FinalWorkerTime[m*19/20])
+		}
+	}
+	if len(result) > 0 {
+		return result + "}"
+	}
```

Comment:
Duplicate code with line 892 ~ 908. Could you do some refactor?

Reply:
Maybe you can add a method for `AggWorkerStat` to avoid duplicate code.

Reply:
good idea